File: leetcode/cpp/15.3sum.cpp
class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        if (nums.size() < 3)
            return {};
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for (int i = 0; i < nums.size() - 2; ++i)
        {
            int j = i + 1;
            int k = nums.size() - 1;
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            while (j < k)
            {
                if (nums[i] + nums[j] + nums[k] == 0)
                {
                    res.push_back({nums[i], nums[j], nums[k]});
                    ++j;
                    while (j < k && nums[j] == nums[j - 1])
                        ++j;
                    --k;
                    while (k > j && nums[k] == nums[k + 1])
                        --k;
                }
                else
                {
                    if (nums[i] + nums[j] + nums[k] > 0)
                    {
                        --k;
                    }
                    else
                    {
                        ++j;
                    }
                }
            }
        }
        return res;
    }
};

File: leetcode/python/15.3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            target = 0 - nums[i]
            start, end = i+1, len(nums)-1
            while start < end:
                if nums[start] + nums[end] > target:
                    end -= 1
                elif nums[start] + nums[end] < target:
                    start += 1
                else:
                    res.append([nums[i], nums[start], nums[end]])
                    end -= 1
                    start += 1
                    while start < end and nums[end] == nums[end+1]:
                        end -= 1
                    while start < end and nums[start] == nums[start-1]:
                        start += 1
        return res
